platform/particle_electron/export/aux/uart_bus.hpp

```diff
@@ -0,0 +1,246 @@
+
+    static safe_storage<handler_fn>     m_ev_storage;    //! Storage for event handler.
+};
+
+//------------------------------------------------------------------------------
+
+template<uart_device dev>
+safe_storage<typename uart_bus<dev>::handler_fn> uart_bus<dev>::m_ev_storage;
+
+template<uart_device dev>
+const uint8_t *uart_bus<dev>::m_tx;
+
+template<uart_device dev>
+uint8_t *uart_bus<dev>::m_rx;
+
+template<uart_device dev>
+size_t uart_bus<dev>::m_tx_size;
+
+template<uart_device dev>
+size_t uart_bus<dev>::m_rx_size;
+
+//------------------------------------------------------------------------------
+
+template<uart_device dev>
+err uart_bus<dev>::init()
+{
+    auto &serial = get_serial();
+
+    serial.begin();
+    m_ev_storage.init();
+
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::set_rx(uint8_t *rx, size_t size)
+{
+    m_rx = rx;
+    m_rx_size = size;
+
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::set_tx(size_t size, uint8_t fill_byte)
+{
+    (void)size;
+    (void)fill_byte;
+    // TODO
+    return err::generic;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::set_tx(const uint8_t *tx, size_t size)
+{
+    m_tx = tx;
+    m_tx_size = size;
+
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::set_handler(const handler_fn &handler)
+{
+    m_ev_storage.get() = handler;
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::reset_buffers()
+{
+    m_tx = nullptr;
+    m_tx_size = 0;
+
+    m_rx = nullptr;
+    m_rx_size = 0;
+
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::reset_handler()
+{
+    m_ev_storage.get() = handler_fn{};
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::do_xfer()
+{
+    auto &serial = get_serial();
+
+    if (m_tx) {
+        serial.write(m_tx, m_tx_size);
+        m_ev_storage.get()(channel::tx, event::tc, m_tx_size);
+    }
+
+    m_ev_storage.get()(channel::meta, event::tc, 0);
+
+    // TODO: RX.
+
+    return err::ok;
+}
+
+template<uart_device dev>
+err uart_bus<dev>::cancel_xfer()
+{
+    // TODO
+    return err::generic;
+}
+
+//------------------------------------------------------------------------------
+
+template<uart_device dev>
+template<uart_device d, typename std::enable_if_t<d == uart_device::serial0, int>>
+constexpr USBSerial &uart_bus<dev>::get_serial()
+{
+    switch (dev) {
+    case uart_device::serial0:
+        return Serial;
+    }
+}
+
+template<uart_device dev>
+template<uart_device d, typename std::enable_if_t<d != uart_device::serial0, int>>
+constexpr USARTSerial &uart_bus<dev>::get_serial()
+{
+    switch (dev) {
+    case uart_device::serial1:
+        return Serial1;
+    case uart_device::serial2:
+        return Serial2;
+    case uart_device::serial4:
+        return Serial4;
+    case uart_device::serial5:
+        return Serial5;
+    }
```

Comment:
No needed in case of `constexpr` - possible cases are known in compile-time. If invalid input parameter will be given then build will fail.